include/psen_scan_v2/scanner_v2.h

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2020 Pilz GmbH & Co. KG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#ifndef PSEN_SCAN_V2_SCANNER_V2_H
+#define PSEN_SCAN_V2_SCANNER_V2_H
+
+#include <stdexcept>
+#include <memory>
+#include <mutex>
+#include <future>
+#include <chrono>
+#include <functional>
+
+#include "psen_scan_v2/scanner_interface.h"
+#include "psen_scan_v2/scanner_events.h"
+#include "psen_scan_v2/scanner_state_machine.h"
+#include "psen_scan_v2/laserscan.h"
+
+#include "psen_scan_v2/udp_client.h"
+#include "psen_scan_v2/raw_scanner_data.h"
+
+#include "psen_scan_v2/watchdog.h"
+
+namespace psen_scan_v2
+{
+using namespace psen_scan_v2::scanner_protocol;
+using std::placeholders::_1;
+using std::placeholders::_2;
+
+// TODO: Move to ScannerController class and read from ScannerConfiguration
+static constexpr unsigned short DATA_PORT_OF_SCANNER_DEVICE{ 2000 };
+static constexpr unsigned short CONTROL_PORT_OF_SCANNER_DEVICE{ 3000 };
+
+static constexpr std::chrono::milliseconds REPLY_TIMEOUT{ 1000 };
+
+class ScannerV2 : public IScanner
+{
+public:
+  ScannerV2(const ScannerConfiguration& scanner_config, const LaserScanCallback& laser_scan_cb);
+  ~ScannerV2();
+
+public:
+  std::future<void> start() override;
+  std::future<void> stop() override;
+
+private:
+  // Raw pointer used here because "msm::back::state_machine" cannot properly pass
+  // a "std::unique_ptr" to "msm::front::state_machine_def".
+  StateMachineArgs* createStateMachineArgs();
+
+  template <class T>
+  void triggerEventWithParam(const T& event);
+
+  template <class T>
+  void triggerEvent();
+
+  void scannerStartedCB();
```

Comment:
As far as I understood the code: If `startStartWatchdog` and `stopStartWatchdog` would be part of the state machine and not of `ScannerV2` the and the future would be passed to the SM using the event (while returning the promise to the user) the scanner would no longer need to pass own callback to the SM and the encapsulation would be stronger... But maybe I am missing something...

Reply:
I think you mean we could pass the `promise` to the SM and the `future` to the user`, right?

Yes, this is definitely a good idea. I give it a try.

Reply:
The problem is that a `promise` cannot be copied. The `msm::front::state_machine_def<...>` tries to make a copy of the event which it gets via `trigger_event`. If the `promise` is part of the event, the event cannot be copied. Therefore, the approach sounds good but is not realizable. (Of course a hack would be possible but, in my opinion, this is not the way to go here.)

Reply:
```
/////////////////////////////////////
// Compile with g++ <file> -lboost_system -std=c++11 -pthread
/////////////////////////////////////

#include <future>
#include <iostream>
#include <string>

// back-end
#include <boost/msm/back/state_machine.hpp>
// front-end
#include <boost/msm/front/state_machine_def.hpp>

namespace msm = boost::msm;
namespace mpl = boost::mpl;

struct sm_ : public msm::front::state_machine_def<sm_> {
  std::promise<std::string> promise_;

  struct eventB {};

  struct states {
    struct A : public msm::front::state<> {
      template <class Event, class FSM> void on_exit(Event const &, FSM &) {
        std::cerr << "Leaving A\n";
      }
    };
    struct B : public msm::front::state<> {
      template <class Event, class FSM> void on_entry(Event const &, FSM &) {
        std::cerr << "Entering B\n";
      }
    };
  };

  void action(eventB const &) {
    try {
      promise_.set_value("new promise value");
    } catch (const std::future_error &e) {
      std::cerr << "Future error" << '\n';
    } catch (const std::exception &e) {
      std::cerr << e.what() << '\n';
    }
  }

  struct transition_table
      : mpl::vector<
            //    Start          Event             Next       Action  Guard
            //  +--------------+-----------------+-----------+------+-------+
            a_row<states::A, eventB, states::B, &sm_::action>
            //  +--------------+-----------------+-----------+------+-------+
            > {};

  typedef states::A initial_state;
};

// Pick a back-end
typedef msm::back::state_machine<sm_> sm;

int main() {
  sm state_maschine;

  std::promise<std::string> promise;
  std::future<std::string> future = promise.get_future();
  state_maschine.promise_ = std::move(promise);

  state_maschine.process_event(sm_::eventB());
  std::cerr << "Value is \"" << future.get() << "\"\n";
}
```

Reply:
Yes, we could move the promise into the the state machine front-end. But what happens if the `Scanner::start()` function  is called again? Then we need to reset the promise like it is done in the `ScannerController`. In my opinion, this is not really a good alternative.